File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

	static int[][] map;
	static int N, sum, min, cnt;
	static boolean[] visited;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		map = new int[N][N];
		visited = new boolean[N];
		for (int i = 0; i < N; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
				sum += map[i][j];
			}
		}
		min = sum;
//		cnt = 0;
		combi(0, 0);
		System.out.println(min);
	}

	private static void combi(int depth, int start) {
		if (depth == N/2) {
			int tot1 = 0;
			for (int i = 0; i < N; i++) {
				for (int j = 0; j < N; j++) {
					if (visited[i] && visited[j])
						tot1 += map[i][j];
				}
			}
			int tot2 = 0;
			for (int i = 0; i < N; i++) {
				for (int j = 0; j < N; j++) {
					if (!visited[i] && !visited[j])
						tot2 += map[i][j];
				}
			}

			min = Math.min(min, Math.abs(tot1 - tot2));
//			System.out.println(Arrays.toString(visited));
//			System.out.println(min);
//			cnt++;
			return;
		}
		for (int i = start; i < N; i++) {
			visited[i] = true;
			combi(depth + 1, i + 1);
			visited[i] = false;
		}

	}

}

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
# 복잡도 20 C 10

N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]
R = N // 2


def combi(depth, start):
    if depth == R:
        ans[0] = min(ans[0], clac())
        return
    for i in range(start, N):
        visited[i] = 1
        combi(depth + 1, i + 1)
        visited[i] = 0


def clac():
    tot0, tot1 = 0, 0
    for i in range(N):
        for j in range(i + 1, N):
            if visited[i] == visited[j] == 0:
                tot0 += arr[i][j] + arr[j][i]
            elif visited[i] == visited[j] == 1:
                tot1 += arr[i][j] + arr[j][i]
    return abs(tot0 - tot1)


ans = [float("inf")]
visited = [0] * N
combi(0, 0)
print(ans[0])